src/io/flutter/run/daemon/FlutterAppManager.java

```diff
@@ -300,13 +312,19 @@ private void eventLogMessage(@NotNull AppLog message, @NotNull FlutterDaemonCont
   }
 
   private void eventDeviceAdded(@NotNull DeviceAdded added, @NotNull FlutterDaemonController controller) {
-    myService.addConnectedDevice(new FlutterDevice(added.name, added.id, added.platform, added.emulator));
+    synchronized (myLock) {
+      myService.addConnectedDevice(new FlutterDevice(added.name, added.id, added.platform, added.emulator));
+    }
   }
 
   private void eventDeviceRemoved(@NotNull DeviceRemoved removed, @NotNull FlutterDaemonController controller) {
-    for (ConnectedDevice device : myService.getConnectedDevices()) {
-      if (device.deviceName().equals(removed.name) && device.deviceId().equals(removed.id) && device.platform().equals(removed.platform)) {
-        myService.removeConnectedDevice(device);
+    synchronized (myLock) {
+      for (ConnectedDevice device : myService.getConnectedDevices()) {
+        if (device.deviceName().equals(removed.name) &&
+            device.deviceId().equals(removed.id) &&
```

Comment:
devices should be unique by id - you won't have any on the same machine with the same ID